File: src/sync.ts
/** @fileoverview Coordinates importing libraries, generating resolved rules, and applying complete project changes. */

import { importLibraries } from './imports';
import { buildRules } from './builds';
import { applyChanges, withWriter } from './project-files/apply';
import { vendorFiles } from './project-files/snapshots';
import {
  projectLocation,
  readProject,
  requireActive,
  buildInput,
  generatedBytes,
  assertUnchanged,
  compareFiles,
  managedFiles,
} from './project-files/project';
import type { ProjectOptions, FileChanges } from './project-files/project';
export type { ProjectOptions, FileChanges } from './project-files/project';
export { ProjectError } from './project-files/files';

/** Import every source and generate resolved rules before replacing managed output; preserve authored input and recover interrupted writes. */
export async function sync(options: ProjectOptions = {}): Promise<FileChanges> {
  requireActive(options);
  const { root, configPath } = await projectLocation(options);
  return withWriter(root, async () => {
    requireActive(options);
    const state = await readProject(root, configPath);
    const libraries = await importLibraries(
      state.input,
      options.signal === undefined ? {} : { signal: options.signal },
    );
    const snapshots = Object.fromEntries(
      Object.entries(libraries).map(([alias, library]) => [
        alias,
        library.snapshot,
      ]),
    );
    const generated = generatedBytes(
      buildRules(buildInput(state, snapshots, options)).files,
    );
    const vendor = vendorFiles(libraries);
    const changes = compareFiles(
      managedFiles(
        state.vendor?.files ?? new Map(),
        state.generated?.files ?? new Map(),
      ),
      managedFiles(vendor, generated),
    );
    await applyChanges(root, { vendor, generated }, () =>
      assertUnchanged(root, configPath, state, options),
    );
    return changes;
  });
}

File: internal/project/sync.go
// Fetch libraries and refresh vendor, generated output, and the managed guide under one recoverable writer.

package project

import (
	"context"
	"fmt"

	"github.com/fabricahq/code-rules/internal/build"
	"github.com/fabricahq/code-rules/internal/filetxn"
	"github.com/fabricahq/code-rules/internal/imports"
)

// Sync re-resolves configured revisions, validates and renders everything before replacing managed trees.
// Git settings are trusted caller options. Any failure returns no change report; authored files stay untouched.
func Sync(ctx context.Context, options Options, git imports.Options) (FileChanges, error) {
	if err := ctx.Err(); err != nil {
		return FileChanges{}, err
	}
	root, name, err := projectLocation(options.ConfigPath)
	if err != nil {
		return FileChanges{}, err
	}
	defer root.Close()
	var changes FileChanges
	err = filetxn.WithWriter(ctx, root, func(w *filetxn.Writer) error {
		before, err := readProject(ctx, root, name)
		if err != nil {
			return err
		}
		guide, err := planProjectGuide(ctx, root, name)
		if err != nil {
			return err
		}
		imported, err := imports.ImportLibraries(ctx, before.config, git)
		if err != nil {
			return err
		}
		snapshots := map[string]snapshot{}
		libraries := map[string]build.Library{}
		for alias, item := range imported {
			snapshots[alias] = item.Snapshot
			libraries[alias] = build.Library{Catalog: item.Catalog, Commit: item.Snapshot.Commit, Tag: item.Snapshot.Tag}
		}
		vendor, err := encodeSnapshots(before.config, snapshots)
		if err != nil {
			return err
		}
		output, err := renderProject(ctx, before, libraries, options)
		if err != nil {
			return err
		}
		changes = compareFiles(managedFiles(treeFiles(before.vendor), treeFiles(before.generated)), managedFiles(vendor, output.Files))
		targets := map[filetxn.Target]map[string][]byte{filetxn.Vendor: vendor, filetxn.Generated: output.Files}
		includeProjectGuide(targets, guide, &changes)
		return w.Apply(targets, func() error {
			if err := requireUnchanged(ctx, root, name, before); err != nil {
				return err
			}
			return requireGuideUnchanged(ctx, root, guide)
		})
	})
	if err != nil {
		return FileChanges{}, fmt.Errorf("sync project: %w", err)
	}
	return changes, nil
}

// managedFiles prefixes source-relative and generated-relative paths for an unambiguous combined change report.
func managedFiles(vendor, generated map[string][]byte) map[string][]byte {
	result := make(map[string][]byte, len(vendor)+len(generated))
	for name, data := range vendor {
		result["vendor/"+name] = data
	}
	for name, data := range generated {
		result["generated/"+name] = data
	}
	return result
}
